### PR TITLE
Add Seeed Wio 3G support

### PR DESCRIPTION
**Pull request by toyowata**
The target uses STM32F439VI device.

Internal Flash : 2MB
RAM : 256kB
Connectivity : Cellular

kvstore : internal flash
Update storage :  internal flash (1MB)

### CODE REVIEW

**Comment by iot-pdmc** [comment]
Thanks for your changes.
@ARMmbed/pdmc-integrators please review.

File: mbed_app.json [hunk 1]
```diff
@@ -317,6 +317,34 @@
...
+        },
+        "WIO_3G": {
+            "target.features_add"                       : ["LWIP", "STORAGE", "BOOTLOADER"],
+            "target.macros_add"                         : ["MBEDTLS_USER_CONFIG_FILE=\"mbedTLSConfig_mbedOS.h\""],
+            "target.device_has_remove"                  : ["SERIAL_FC"],
+            "target.network-default-interface-type"     : "CELLULAR",
+            "target.components_add"                     : ["FLASHIAP"],
+            "target.bootloader_img"                     : "bootloader/mbed-bootloader-wio_3g-internal_flash-no_rot-v4.1.2.bin",
+            "target.header_offset"                      : "0x8000",
+            "target.app_offset"                         : "0x8400",
```
Comment (teetak01) [inline mbed_app.json]: target.restrict_size should be added to prevent overflowing to second bank.

File: mbed_app.json [hunk 2]
```diff
@@ -317,6 +317,34 @@
...
+            "lwip.pbuf-pool-size"                       : 16,
+            "lwip.mem-size"                             : 12500,
+            "update-client.bootloader-details"          : "0x08005ee8",
+            "update-client.application-details"         : "(MBED_ROM_START + MBED_BOOTLOADER_SIZE)",
+            "update-client.storage-address"             : "(MBED_CONF_STORAGE_TDB_INTERNAL_INTERNAL_BASE_ADDRESS+MBED_CONF_STORAGE_TDB_INTERNAL_INTERNAL_SIZE)",
+            "update-client.storage-size"                : "(1024*1024-MBED_CONF_STORAGE_TDB_INTERNAL_INTERNAL_SIZE)",
+            "update-client.storage-page"                : 1,
+            "mbed-cloud-client.update-storage"          : "ARM_UCP_FLASHIAP",
+            "storage_tdb_internal.internal_base_address": "(MBED_ROM_START+1024*1024)",
+            "storage_tdb_internal.internal_size"        : "(128*1024)",
```
Comment (teetak01) [inline mbed_app.json]: 32 KiB would be enough.

Comment (teetak01) [reply mbed_app.json]: also makes nice symmetry with the bootloader-size.

**Comment by teetak01** [comment]
build-only

**Comment by iot-pdmc** [comment]
# [Test run: SUCCESS](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/532/)
***
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/532)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by toyowata** [comment]
## Test result

```
(venv3) 11:59:36 mbed-os-example-pelion-1 $ pytest TESTS/pelion-e2e-python-test-library/tests/dev-client-tests.py
=================================================== test session starts ===================================================
platform darwin -- Python 3.8.1, pytest-5.3.1, py-1.8.0, pluggy-0.13.1
rootdir: /Users/toywat01/test/seeed/wio/mbed-os-example-pelion-1/TESTS/pelion-e2e-python-test-library/tests, inifile: pytest.ini
plugins: html-2.0.1, metadata-1.8.0
collected 4 items                                                                                                         

TESTS/pelion-e2e-python-test-library/tests/dev-client-tests.py::test_01_get_device_id 
----------------------------------------------------- live log setup ------------------------------------------------------
2020-01-10 11:59:41.311      INFO   Using "WIO_3G[0]: 901400001fe2140000000000000000000000000097969902" device at "/dev/tty.usbmodem14202" port for tests
2020-01-10 11:59:43.316      INFO   Starting runner threads for client "D0"
2020-01-10 11:59:43.751      INFO   <--|D0| 
2020-01-10 11:59:43.753      INFO   <--|D0| Mbed Bootloader
2020-01-10 11:59:43.959      INFO   <--|D0| No Update image
2020-01-10 11:59:43.962      INFO   <--|D0| [DBG ] Active firmware up-to-date
2020-01-10 11:59:43.963      INFO   <--|D0| booting...
2020-01-10 11:59:43.964      INFO   <--|D0| 
2020-01-10 11:59:43.968      INFO   <--|D0| Application ready
2020-01-10 11:59:44.199      INFO   <--|D0| Connect to network
2020-01-10 12:00:05.677      INFO   <--|D0| Network initialized, connected with IP 10.129.102.246
2020-01-10 12:00:05.678      INFO   <--|D0| 
2020-01-10 12:00:05.679      INFO   <--|D0| Start developer flow
2020-01-10 12:00:05.774      INFO   <--|D0| Create resources
2020-01-10 12:00:05.778      INFO   <--|D0| Register Pelion Device Management Client
2020-01-10 12:00:05.778      INFO   <--|D0| 
2020-01-10 12:00:10.795      INFO   <--|D0| Counter 1
2020-01-10 12:00:15.800      INFO   <--|D0| Counter 2
2020-01-10 12:00:20.806      INFO   <--|D0| Counter 3
2020-01-10 12:00:23.416      INFO   <--|D0| Client registered.
2020-01-10 12:00:23.420      INFO   <--|D0| Account ID: 016008ff3058be206ae5ecf600000000
2020-01-10 12:00:23.424      INFO   <--|D0| Endpoint name: 016f8d652531000000000001001ecff2
2020-01-10 12:00:23.428      INFO   <--|D0| Device ID: 016f8d652531000000000001001ecff2
2020-01-10 12:00:23.429      INFO   <--|D0| 
------------------------------------------------------ live log call ------------------------------------------------------
2020-01-10 12:00:23.431      INFO   Device registered to cloud with device ID: 016f8d652531000000000001001ecff2
PASSED                                                                                                              [ 25%]
TESTS/pelion-e2e-python-test-library/tests/dev-client-tests.py::test_02_put_resource 
----------------------------------------------------- live log setup ------------------------------------------------------
2020-01-10 12:00:25.029      INFO   Created new developer api key for test case, id: 016f8d6550dd8a6f242a476400000000
2020-01-10 12:00:25.029      INFO   Register and open WebSocket notification channel
2020-01-10 12:00:25.811      INFO   <--|D0| Counter 4
2020-01-10 12:00:30.817      INFO   <--|D0| Counter 5
2020-01-10 12:00:30.846      INFO   Opening WebSocket handler
2020-01-10 12:00:30.847      INFO   Starting WebSocket threads
------------------------------------------------------ live log call ------------------------------------------------------
2020-01-10 12:00:31.662      INFO   WebSocket opened to wss://api.us-east-1.mbedcloud.com/v2/notification/websocket-connect
2020-01-10 12:00:31.676      INFO   Put value "1:2:3:4" to resource "/3201/0/5853". Async-id: "2345099a-b7e6-492d-a8b1-144f6e55869b"
2020-01-10 12:00:31.721      INFO   <--|D0| PUT update 0
2020-01-10 12:00:35.681      INFO   Async response received: "b''"
PASSED                                                                                                              [ 50%]
TESTS/pelion-e2e-python-test-library/tests/dev-client-tests.py::test_03_get_resource 
------------------------------------------------------ live log call ------------------------------------------------------
2020-01-10 12:00:35.822      INFO   <--|D0| Counter 6
2020-01-10 12:00:36.525      INFO   Get value from resource "/3201/0/5853". Async-id: "a970bc36-2641-478f-ad32-9844d15a81c4"
2020-01-10 12:00:37.526      INFO   Async response received: "b'1:2:3:4'"
PASSED                                                                                                              [ 75%]
TESTS/pelion-e2e-python-test-library/tests/dev-client-tests.py::test_04_subscribe_resource 
------------------------------------------------------ live log call ------------------------------------------------------
2020-01-10 12:00:38.450      INFO   Set subscription for resource "/3201/0/5853". Async-id: "016f8d65-8534-2609-4a57-3b1100000000"
2020-01-10 12:00:39.367      INFO   Put value "500:500:500:500" to resource "/3201/0/5853". Async-id: "b1a76349-80ab-4c6b-9628-26df5e81060b"
2020-01-10 12:00:39.401      INFO   <--|D0| PUT update 0
2020-01-10 12:00:40.370      INFO   Notification value received: "500:500:500:500"
PASSED                                                                                                              [100%]
---------------------------------------------------- live log teardown ----------------------------------------------------
2020-01-10 12:00:40.372      INFO   Closing WebSocket threads
2020-01-10 12:00:40.827      INFO   <--|D0| Counter 7
2020-01-10 12:00:42.372      INFO   Deleting WebSocket channel
2020-01-10 12:00:43.317      INFO   Cleaning out the generated test case developer api key, id: 016f8d6550dd8a6f242a476400000000
2020-01-10 12:00:44.179      INFO   Closing client "016f8d652531000000000001001ecff2"
2020-01-10 12:00:45.833      INFO   <--|D0| Counter 8
------------------------------------------------- live log sessionfinish --------------------------------------------------
2020-01-10 12:00:47.038      INFO   -----  TEST RESULTS SUMMARY  -----
2020-01-10 12:00:47.039      INFO   [ check the complete fail reasons and code locations from this log or html report ]
2020-01-10 12:00:47.039      INFO   [passed] - dev-client-tests.py::test_01_get_device_id - (0.808s)
2020-01-10 12:00:47.039      INFO   [passed] - dev-client-tests.py::test_02_put_resource - (4.833s)
2020-01-10 12:00:47.039      INFO   [passed] - dev-client-tests.py::test_03_get_resource - (1.842s)
2020-01-10 12:00:47.039      INFO   [passed] - dev-client-tests.py::test_04_subscribe_resource - (2.840s)


========================= 4 passed in 67.30s (0:01:07) =========================
2020-01-10 12:00:47,038:tests.conftest:MainThread:INFO: -----  TEST RESULTS SUMMARY  -----
2020-01-10 12:00:47,039:tests.conftest:MainThread:INFO: [ check the complete fail reasons and code locations from this log or html report ]
2020-01-10 12:00:47,039:tests.conftest:MainThread:INFO: [passed] - dev-client-tests.py::test_01_get_device_id - (0.808s)
2020-01-10 12:00:47,039:tests.conftest:MainThread:INFO: [passed] - dev-client-tests.py::test_02_put_resource - (4.833s)
2020-01-10 12:00:47,039:tests.conftest:MainThread:INFO: [passed] - dev-client-tests.py::test_03_get_resource - (1.842s)
2020-01-10 12:00:47,039:tests.conftest:MainThread:INFO: [passed] - dev-client-tests.py::test_04_subscribe_resource - (2.840s)
```

File: mbed_app.json (hunk 1 above)
Comment (toyowata) [reply mbed_app.json]: Thanks.  I have added target.restrict_size for this.

File: mbed_app.json (hunk 2 above)
Comment (toyowata) [reply mbed_app.json]: You are right.  `storage_tdb_internal.internal_size` should be 32kB (16*2) for the target.  I also updated the bootloader binary.

**Comment by yogpan01** [comment]
@teetak01 Is this PR ready to be merged?
@MarceloSalazar FYI

**Comment by teetak01** [comment]
build-only

**Comment by iot-pdmc** [comment]
# [Test run: SUCCESS](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/533/)
***
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/533)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by MarceloSalazar** [comment]
Note WIO_3G platform is in Raas (helga).

**Comment by teetak01** [comment]
build-only

**Comment by teetak01** [comment]
build-only

**Comment by iot-pdmc** [comment]
# [Test run: SUCCESS](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/555/)
***
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/555)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by iot-pdmc** [comment]
# [Test run: ABORTED](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/556/)
posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by teetak01** [comment]
build-only

**Comment by iot-pdmc** [comment]
# [Test run: SUCCESS](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/569/)
***
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/569)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Comment by toyowata** [comment]
@teetak01 
rebased the PR.

**Comment by teetak01** [comment]
build-only

**Comment by iot-pdmc** [comment]
# [Test run: SUCCESS](https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/579/)
***
[Build artifacts](http://pdmc-ci.s3-website-eu-west-1.amazonaws.com/?prefix=jenkins-ci/mbed-os-example-pelion/579)

posted-by-tag: https://jenkins-client.isgtesting.com/job/mbed-os-example-pelion/

**Pull request merged**